waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -0,0 +1,168 @@
+  import libp2p/debugutils
+
+logScope:
+  topics = "nim-waku noise"
+
+const
+  # Empty is a special value which indicates k has not yet been initialized.
+  EmptyKey = default(ChaChaPolyKey)
+  NonceMax = uint64.high - 1 # max is reserved
+  NoiseSize = 32
+  MaxPlainSize = int(uint16.high - NoiseSize - ChaChaPolyTag.len)
+
+
+type
+  KeyPair* = object
+    privateKey: Curve25519Key
+    publicKey: Curve25519Key
+
+  NoisePublicKey* = object
+    flag: uint8
+    pk: seq[byte]
+
+  ChaChaPolyCiphertext* = object
+    data: seq[byte]
+    tag: ChaChaPolyTag
+
+  ChaChaPolyCipherState* = object
+    k*: ChaChaPolyKey
+    nonce*: ChaChaPolyNonce
+    ad*: seq[byte]
+
+  NoiseError* = object of LPError
+  NoiseHandshakeError* = object of NoiseError
+  NoiseDecryptTagError* = object of NoiseError
+  NoiseNonceMaxError* = object of NoiseError # drop connection on purpose
+  NoisePublicKeyError* = object of NoiseError
+  NoiseMalformedHandshake* = object of NoiseError
+
+
+
+#################################################################
+
+
+# ChaChaPoly encryption
+proc encrypt*(
+    state: ChaChaPolyCipherState,
+    plaintext: openArray[byte]): ChaChaPolyCiphertext
+    {.noinit, raises: [Defect].} =
+  #TODO: add padding
+  result.data.add plaintext
+  ChaChaPoly.encrypt(state.k, state.nonce, result.tag, result.data, state.ad)
+
+proc decrypt*(
+    state: ChaChaPolyCipherState, 
+    ciphertext: ChaChaPolyCiphertext): seq[byte]
+    {.raises: [Defect, NoiseDecryptTagError].} =
+  var
+    tagIn = ciphertext.tag
+    tagOut: ChaChaPolyTag
+  result = ciphertext.data
+  ChaChaPoly.decrypt(state.k, state.nonce, tagOut, result, state.ad)
+  #TODO: add unpadding
+  trace "decrypt", tagIn = tagIn.shortLog, tagOut = tagOut.shortLog, nonce = state.nonce
+  if tagIn != tagOut:
+    debug "decrypt failed", result = shortLog(result)
+    raise newException(NoiseDecryptTagError, "decrypt tag authentication failed.")
+
+
+proc randomChaChaPolyCipherState*(rng: var BrHmacDrbgContext): ChaChaPolyCipherState =
+  brHmacDrbgGenerate(rng, result.k)
+  brHmacDrbgGenerate(rng, result.nonce)
+  result.ad = newSeq[byte](32)
+  brHmacDrbgGenerate(rng, result.ad)
+
+
+#################################################################
+
+
+# Utility
+
+proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
+  result.privateKey = Curve25519Key.random(rng)
+  result.publicKey = result.privateKey.public()
+
+
+# Public keys serializations/encryption
+
+proc `==`(k1, k2: NoisePublicKey): bool =
+  result = (k1.flag == k2.flag) and (k1.pk == k2.pk)
+  
+
+proc keyPairToNoisePublicKey*(keyPair: KeyPair): NoisePublicKey =
+  result.flag = 0
+  result.pk = getBytes(keyPair.publicKey)
+
+
+proc genNoisePublicKey*(rng: var BrHmacDrbgContext): NoisePublicKey =
+  let keyPair: KeyPair = genKeyPair(rng)
+  result.flag = 0
+  result.pk = getBytes(keyPair.publicKey)
+
+proc serializeNoisePublicKey*(noisePublicKey: NoisePublicKey): seq[byte] =
+  result.add noisePublicKey.flag
```

Comment:
Would be good to explain the serialization pattern in a comment.

Reply:
Addressed in https://github.com/status-im/nim-waku/commit/704f3fe40fb4ad63212ec5f6face687b4971c909